core/controllers/admin_test.py

```diff
@@ -1109,18 +1109,139 @@ def test_send_dummy_mail(self):
+            payload={
+                'old_username': self.OLD_USERNAME,
+                'new_username': 123},
+            csrf_token=csrf_token,
+            expected_status_int=400)
+        self.assertEqual(
+            response['error'], 'Expected new username to be a unicode '
+            'string, received 123')
+
+    def test_update_username_with_non_string_old_username(self):
+        csrf_token = self.get_new_csrf_token()
+
+        response = self.put_json(
+            '/updateusernamehandler',
+            payload={
+                'old_username': 123,
+                'new_username': self.NEW_USERNAME},
+            csrf_token=csrf_token,
+            expected_status_int=400)
+        self.assertEqual(
+            response['error'], 'Expected old username to be a unicode '
+            'string, received 123')
+
+    def test_update_username_with_long_new_username(self):
+        long_username = 'a' * (constants.MAX_USERNAME_LENGTH + 1)
+        csrf_token = self.get_new_csrf_token()
+
+        response = self.put_json(
+            '/updateusernamehandler',
+            payload={
+                'old_username': self.OLD_USERNAME,
+                'new_username': long_username},
+            csrf_token=csrf_token,
+            expected_status_int=400)
+        self.assertEqual(
+            response['error'], 'Expected new username to be less than %s '
+            'characters, received %s' % (
+                constants.MAX_USERNAME_LENGTH,
+                long_username))
+
+    def test_update_username_with_nonexistent_old_username(self):
+        non_existent_username = 'invalid'
+        csrf_token = self.get_new_csrf_token()
+
+        response = self.put_json(
+            '/updateusernamehandler',
+            payload={
+                'old_username': non_existent_username,
+                'new_username': self.NEW_USERNAME},
+            csrf_token=csrf_token,
+            expected_status_int=400)
+        self.assertEqual(response['error'], 'Invalid username: invalid')
+
+    def test_update_username_with_new_username_already_taken(self):
+        csrf_token = self.get_new_csrf_token()
+
+        response = self.put_json(
+            '/updateusernamehandler',
+            payload={
+                'old_username': self.OLD_USERNAME,
+                'new_username': self.OLD_USERNAME},
+            csrf_token=csrf_token,
+            expected_status_int=400)
+        self.assertEqual(response['error'], 'Username already taken.')
+
+    def test_update_username(self):
+        user_id = self.get_user_id_from_email(self.ADMIN_EMAIL)
+        csrf_token = self.get_new_csrf_token()
+
+        self.put_json(
+            '/updateusernamehandler',
+            payload={
+                'old_username': self.OLD_USERNAME,
+                'new_username': self.NEW_USERNAME},
+            csrf_token=csrf_token)
+        self.assertEqual(user_services.get_username(user_id), self.NEW_USERNAME)
```

Comment:
Do we also need to check the existence of the new audit model?

Reply:
@DubeySandeep Don't the tests present [here](https://github.com/oppia/oppia/pull/9191/files#diff-c18d36062f4b6168e611a66d20f3077eR90), cover this case?

Reply:
Ah, that's a unit test for the Model, I'm asking for a test which ensures that changing username creates audit model. [More around integration test]

Reply:
Thanks, I understand now. I've added a test.